features/310-prerender-gear/prerender.md

```diff
@@ -0,0 +1,105 @@
+
+Prerender handles any incoming request by first checking whether there is any cached DOM for the forwarded url. It there is a cache hit, prerender "hits the fast path" and returns the cached DOM as-is. Otherwise it proceeds to the slow path.
+    
+The slow path begins by first navigating to the forwarded url using a headless browser, aloowing the browser to proceed according to redirect transparently, then inspect the final response once it settles down to check if the url locates an html document (i.e. `content-type` is `text/html`). If not, prerender returns the response body and headers directly, otherwise it proceeds to capture the response body (the html DOM) and the response header.
+
+To allow scripts to run to its completion or to allow DOM to transform to a specific state, prerender waits for a configurable period of time before capturing the resultant DOM. The resultant DOM will have all of its script tags removed, and a `<base>` tag is inserted to the DOM appropriately to fix relative links.
+
+Response headers are then modified. Specifically, `content-length` and `content-encoding` will be removed.
+
+Finally, the modified DOM and headers are put into prerendered html cache using the forwarded url as keys before being sent back to gateway. A configurable expiry time is also applied to the cache such that they expire automatically.
+
+#### Redirect
+
+Note that since the forwarded url is used directly as the cache key, even if the forwarded url in a future request is the same as any of the redirected urls prerender has encountered during navigation, the request is a cache miss. For example:
+
+1. gateway requests url `http://a.com`
+2. prerender sees a caches miss, proceed to navigates to `http://a.com`
+3. prerender is redirected to `http://b.com`
+4. prerender captures the DOM of `http://b.com`, then stores it with key `http://a.com` and serves it
+5. gateway requests url `http://b.com`
+6. prerender sees a cache miss (it only has cache for `http://a.com`), proceeds to navigates to `http://b.com`
+7. prerender captures the DOM of `http://b.com` (again), then stores it with key `http://b.com` and serves it
+
+Notice how in (6) prerender consider `http://b.com` a cache miss even though it has been redirected to it.
+
+### Invalidating cache
+
+Since prerender supports some configuration about cache, when a configuration is changed such that according to the new configuration old cache is no longer valid, it is necessary to manually reset prerendered html cache. Prerender supports invalidation via a `origin` and `path` protocol.
+
+## API
+
+### Configuration
+
+`seo-prerender`: Boolean. `true` to enable prerender, `false` to disable.
+
+`seo-prerender-expiry`: Time. The amount of time a prerendrered html is kept in cache.
```

Comment:
Specify the unit we are using here. Second should be the right unit we want to use.

Reply:
I thought skygear has a universal time format parsing policy that supports things like `00h00m00s`? Since this config is inside app deployment configuration I didn't want to say too much here. Can copy and paste if someone can provide one.

Reply:
I am sorry that we do not have such format at the moment. Ideally we should support literal like `1s` `2m` `3h` to indicate 1 second, 2 minute and 3 hours respectively, just like nginx. So for now just specify it is in the unit of seconds.